src/Multicall3.sol

```diff
@@ -0,0 +1,108 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.12;
+
+/// @title Multicall3
+/// @notice Aggregate results from multiple read-only function calls
+/// @dev Multicall & Multicall2 backwards-compatible
+/// @author Michael Elliot <mike@makerdao.com>
+/// @author Joshua Levine <joshua@makerdao.com>
+/// @author Nick Johnson <arachnid@notdot.net>
+/// @author Andreas Bigger <andreas@nascent.xyz>
+
+contract Multicall3 {
+    struct Call {
+        address target;
+        bytes callData;
+    }
+
+    struct Call3 {
+        address target;
+        bool allowFailure;
+        bytes callData;
+    }
+
+    struct Result {
+        bool success;
+        bytes returnData;
+    }
+
+    function aggregate(Call[] calldata calls) public returns (uint256 blockNumber, bytes[] memory returnData) {
+        blockNumber = block.number;
+        uint256 length = calls.length;
+        returnData = new bytes[](length);
+        for (uint256 i = 0; i < length;) {
+            (bool success, bytes memory ret) = calls[i].target.call(calls[i].callData);
+            require(success, "Multicall3: aggregate failed");
+            returnData[i] = ret;
+            unchecked { ++i; }
+        }
+    }
+
+    function tryAggregate(bool requireSuccess, Call[] calldata calls) public returns (Result[] memory returnData) {
+        uint256 length = calls.length;
+        returnData = new Result[](length);
+        for (uint256 i = 0; i < length;) {
+            (bool success, bytes memory ret) = calls[i].target.call(calls[i].callData);
+            if (requireSuccess) require(success, "Multicall3: tryAggregate failed");
+            returnData[i] = Result(success, ret);
+            unchecked { ++i; }
+        }
+    }
+
+    function tryBlockAndAggregate(bool requireSuccess, Call[] calldata calls) public returns (uint256 blockNumber, bytes32 blockHash, Result[] memory returnData) {
+        blockNumber = block.number;
+        blockHash = blockhash(block.number);
+        returnData = tryAggregate(requireSuccess, calls);
+    }
+
+    function blockAndAggregate(Call[] calldata calls) public returns (uint256 blockNumber, bytes32 blockHash, Result[] memory returnData) {
+        (blockNumber, blockHash, returnData) = tryBlockAndAggregate(true, calls);
+    }
+
+    function aggregate3(Call3[] calldata calls) public returns (uint256 blockNumber, bytes32 blockHash, Result[] memory returnData) {
```

Comment:
Should we change this to only return `Result[] memory returnData`, instead of always returning `blockNumber` and `blockHash`? The rationale for doing so is that instead of being forced to return block number and hash, you can choose which block parameters to return (e.g. maybe you only care about timestamp) by simply including them as calls in the Multicall

Reply:
Leaning towards removing `blockNumber` and `blockHash` although it would follow Multicall and Multicall2 canonical patterns.

For example, snapshot.js ignores the blockNumber for `multicall` https://github.com/abigger87/snapshot.js/blob/master/src/utils.ts#L122

Reply:
I'd be ok with removing them also, and IIRC from Twitter I think @WyseNynja also supports removing them

Curious to hear arguments in favor of keeping these parameters as default return values. The main thing I can think of is that you probably always want at least one of block number, timestamp, or hash, and it's cheaper to include them by default compared to the cost of fetching them inside the `for` loop. But that cost difference is small compared to Geth's [default](https://github.com/ethereum/go-ethereum/blob/3038e480f5297b0e80196c156d9f9b45fd86a0bf/eth/ethconfig/config.go#L92) `RpcGasCap` of 50M and Infura's [default](https://docs.infura.io/infura/networks/ethereum/json-rpc-methods/eth_call) of 300M

Reply:
Let's remove then; I think it makes sense to use exclusion by default here.
We should only be adding return values if necessary.